File: integration_tests/php-pdo-mysql/test_integration.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/generated/queries.php';

function get_mysql_url(): string
{
    $url = getenv('MYSQL_URL');
    if ($url === false || $url === '') {
        fwrite(STDERR, "ERROR: MYSQL_URL environment variable is not set\n");
        exit(1);
    }
    return $url;
}

function parse_mysql_url(string $url): array
{
    $parts = parse_url($url);
    if ($parts === false) {
        fwrite(STDERR, "ERROR: Invalid MYSQL_URL format\n");
        exit(1);
    }
    return [
        'host' => $parts['host'] ?? 'localhost',
        'port' => $parts['port'] ?? 3306,
        'dbname' => ltrim($parts['path'] ?? '/scythe_test', '/'),
        'user' => $parts['user'] ?? 'root',
        'password' => $parts['pass'] ?? '',
    ];
}

function create_pdo(string $url): PDO
{
    $params = parse_mysql_url($url);
    $dsn = sprintf(
        'mysql:host=%s;port=%d;dbname=%s',
        $params['host'],
        $params['port'],
        $params['dbname']
    );
    $pdo = new PDO($dsn, $params['user'], $params['password'], [
        PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
        PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
    ]);
    return $pdo;
}

function setup_schema(PDO $pdo): void
{
    $pdo->exec("DROP TABLE IF EXISTS user_tags");
    $pdo->exec("DROP TABLE IF EXISTS tags");
    $pdo->exec("DROP TABLE IF EXISTS orders");
    $pdo->exec("DROP TABLE IF EXISTS users");
    $schema_path = __DIR__ . '/../sql/mysql/schema.sql';
    $schema_sql = file_get_contents($schema_path);
    if ($schema_sql === false) {
        throw new RuntimeException("Failed to read schema file: {$schema_path}");
    }
    // MySQL requires executing statements one at a time
    $statements = array_filter(array_map('trim', explode(';', $schema_sql)));
    foreach ($statements as $statement) {
        if ($statement !== '') {
            $pdo->exec($statement);
        }
    }
}

function assert_equal(mixed $expected, mixed $actual, string $message): void
{
    if ($expected !== $actual) {
        throw new RuntimeException(
            "Assertion failed: {$message} (expected " . var_export($expected, true)
            . ", got " . var_export($actual, true) . ")"
        );
    }
}

function assert_not_null(mixed $value, string $message): void
{
    if ($value === null) {
        throw new RuntimeException("Assertion failed: {$message} (got null)");
    }
}

function assert_true(bool $value, string $message): void
{
    if (!$value) {
        throw new RuntimeException("Assertion failed: {$message}");
    }
}

function test_create_user(PDO $pdo): int
{
    createUser($pdo, "Alice", "alice@example.com", UsersStatus::active);
    $user = getLastInsertUser($pdo);
    assert_not_null($user, "CreateUser: getLastInsertUser returned null");
    assert_equal("Alice", $user->name, "CreateUser name");
    assert_equal("alice@example.com", $user->email, "CreateUser email");
    echo "PASS: CreateUser\n";
    return $user->id;
}

function test_get_user_by_id(PDO $pdo, int $user_id): void
{
    $user = getUserById($pdo, $user_id);
    assert_not_null($user, "GetUserById returned null for id={$user_id}");
    assert_equal("Alice", $user->name, "GetUserById name");
    assert_equal($user_id, $user->id, "GetUserById id");
    echo "PASS: GetUserById\n";
}

function test_list_active_users(PDO $pdo): void
{
    $users = listActiveUsers($pdo, UsersStatus::active);
    assert_true(count($users) >= 1, "Expected at least 1 active user, got " . count($users));
    $names = array_map(fn($u) => $u->name, $users);
    assert_true(in_array("Alice", $names, true), "Expected 'Alice' in active users");
    echo "PASS: ListActiveUsers\n";
}

function test_create_order(PDO $pdo, int $user_id): int
{
    createOrder($pdo, $user_id, "49.99", "Test order");
    $order = getLastInsertOrder($pdo);
    assert_not_null($order, "CreateOrder: getLastInsertOrder returned null");
    assert_equal($user_id, $order->user_id, "CreateOrder user_id");
    assert_equal("Test order", $order->notes, "CreateOrder notes");
    echo "PASS: CreateOrder\n";
    return $order->id;
}

function test_get_orders_by_user(PDO $pdo, int $user_id): void
{
    $orders = getOrdersByUser($pdo, $user_id);
    assert_true(count($orders) >= 1, "Expected at least 1 order, got " . count($orders));
    assert_equal("Test order", $orders[0]->notes, "GetOrdersByUser notes");
    echo "PASS: GetOrdersByUser\n";
}

function test_delete_user(PDO $pdo, int $user_id): void
{
    // Delete orders first due to FK constraint
    deleteOrdersByUser($pdo, $user_id);
    deleteUser($pdo, $user_id);
    $user = getUserById($pdo, $user_id);
    assert_true($user === null || $user === false, "Expected user to be deleted, but it still exists");
    echo "PASS: DeleteUser\n";
}

try {
    $mysql_url = get_mysql_url();
    $pdo = create_pdo($mysql_url);

    setup_schema($pdo);

    $user_id = test_create_user($pdo);
    test_get_user_by_id($pdo, $user_id);
    test_list_active_users($pdo);
    $order_id = test_create_order($pdo, $user_id);
    test_get_orders_by_user($pdo, $user_id);
    test_delete_user($pdo, $user_id);

    echo "\nALL TESTS PASSED\n";
    exit(0);
} catch (Throwable $e) {
    fwrite(STDERR, "FAIL: " . $e->getMessage() . "\n");
    fwrite(STDERR, $e->getTraceAsString() . "\n");
    exit(1);
}

File: integration_tests/ruby-sqlite3/test_integration.rb
# frozen_string_literal: true

require "sqlite3"
require_relative "generated/queries"

SCHEMA_PATH = File.join(__dir__, "..", "sql", "sqlite", "schema.sql")

def get_database_path
  path = ENV["DATABASE_PATH"]
  if path.nil? || path.empty?
    warn "ERROR: DATABASE_PATH environment variable is not set (use :memory: for in-memory)"
    exit 1
  end
  path
end

def setup_schema(db)
  schema_sql = File.read(SCHEMA_PATH)
  db.execute_batch(schema_sql)
end

def assert_equal(expected, actual, message)
  return if expected == actual

  raise "Assertion failed: #{message} (expected #{expected.inspect}, got #{actual.inspect})"
end

def assert_not_nil(value, message)
  return unless value.nil?

  raise "Assertion failed: #{message} (got nil)"
end

def assert_true(value, message)
  return if value

  raise "Assertion failed: #{message}"
end

def test_create_user(db)
  Queries.create_user(db, "Alice", "alice@example.com", "active")
  user_id = db.last_insert_row_id
  user = Queries.get_user_by_id(db, user_id)
  assert_not_nil(user, "get_user_by_id returned nil after create")
  assert_equal("Alice", user.name, "create_user name")
  assert_equal("alice@example.com", user.email, "create_user email")
  assert_equal("active", user.status, "create_user status")
  assert_true(user.id.positive?, "create_user id should be positive")
  puts "PASS: CreateUser"
  user.id
end

def test_get_user_by_id(db, user_id)
  user = Queries.get_user_by_id(db, user_id)
  assert_not_nil(user, "get_user_by_id returned nil for id=#{user_id}")
  assert_equal("Alice", user.name, "get_user_by_id name")
  assert_equal(user_id, user.id, "get_user_by_id id")
  assert_equal("alice@example.com", user.email, "get_user_by_id email")
  assert_equal("active", user.status, "get_user_by_id status")
  puts "PASS: GetUserById"
end

def test_list_active_users(db)
  users = Queries.list_active_users(db, "active")
  assert_true(users.length >= 1, "Expected at least 1 active user, got #{users.length}")
  names = users.map(&:name)
  assert_true(names.include?("Alice"), "Expected 'Alice' in active users, got #{names}")
  puts "PASS: ListActiveUsers"
end

def test_update_user_email(db, user_id)
  Queries.update_user_email(db, "alice-new@example.com", user_id)
  user = Queries.get_user_by_id(db, user_id)
  assert_not_nil(user, "user not found after update")
  assert_equal("alice-new@example.com", user.email, "update_user_email email")
  puts "PASS: UpdateUserEmail"
end

def test_create_order(db, user_id)
  Queries.create_order(db, user_id, "49.99", "Test order")
  order_id = db.last_insert_row_id
  orders = Queries.get_orders_by_user(db, user_id)
  assert_true(orders.length >= 1, "Expected at least 1 order")
  assert_equal("Test order", orders[0].notes, "create_order notes")
  puts "PASS: CreateOrder"
  order_id
end

def test_get_orders_by_user(db, user_id)
  orders = Queries.get_orders_by_user(db, user_id)
  assert_true(orders.length >= 1, "Expected at least 1 order, got #{orders.length}")
  assert_equal("Test order", orders[0].notes, "get_orders_by_user notes")
  puts "PASS: GetOrdersByUser"
end

def test_get_order_total(db, user_id)
  result = Queries.get_order_total(db, user_id)
  assert_not_nil(result, "get_order_total returned nil")
  puts "PASS: GetOrderTotal"
end

def test_search_users(db)
  results = Queries.search_users(db, "%Ali%")
  assert_true(results.length >= 1, "Expected at least 1 search result, got #{results.length}")
  names = results.map(&:name)
  assert_true(names.include?("Alice"), "Expected 'Alice' in search results, got #{names}")
  puts "PASS: SearchUsers"
end

def test_delete_user(db, user_id)
  deleted_count = Queries.delete_orders_by_user(db, user_id)
  assert_equal(1, deleted_count, "delete_orders_by_user count")
  Queries.delete_user(db, user_id)
  user = Queries.get_user_by_id(db, user_id)
  assert_true(user.nil?, "Expected user to be deleted, but it still exists")
  puts "PASS: DeleteUser"
end

begin
  database_path = get_database_path
  db = SQLite3::Database.new(database_path)

  setup_schema(db)

  user_id = test_create_user(db)
  test_get_user_by_id(db, user_id)
  test_list_active_users(db)
  test_update_user_email(db, user_id)
  order_id = test_create_order(db, user_id)
  test_get_orders_by_user(db, user_id)
  test_get_order_total(db, user_id)
  test_search_users(db)
  test_delete_user(db, user_id)

  puts "\nALL TESTS PASSED"
rescue StandardError => e
  warn "FAIL: #{e.message}"
  warn e.backtrace.first(5).join("\n")
  exit 1
ensure
  db&.close
end
